src/main/c/Makefile

```diff
@@ -42,25 +49,28 @@ ifndef CXX
   endif
 endif
 
-# append existing flags so they override our flags
-CXXFLAGS += -O3 -march=native $(CXXSTD) -Ilibsimdpp-2.0-rc2 -Wall -Werror
+# Ignore the user's CXXFLAGS
```

Comment:
this seems rude to our users, no?  How do they provide system-specific flags?

Reply:
I've had painful experiences in past jobs with different users building the same sources and 
getting different behavior, due to differences in their environment settings.  I rewrote the 
build system at Mentor Graphics to wipe the environment clean before doing anything.

If the user *really* wants different flags, then hey, it's all open-source so they can edit the
Makefile themselves. But then obviously they're responsible for whatever happens. Otherwise, I want us to stick close to a vanilla toolset and options,
because we have to keep the libhail.so compatible with the dynamically-generated code,
and adding arbitrary compiler options is asking for trouble.

As discussed before, my preferred solution is actually to lock things down completely by
integrating one particular compiler (probably latest stable clang) into the hail release, because
that way you can guarantee that users are running the same software you tested.  Locking
down the compiler options is a very tiny step in that direction.

Reply:
Seems rude to me.  In particular, I'm going to expect to override these things like I would with any other build system and then have to dig around to figure out why my overrides are being ignored.  I think I had a related problem in another branch of yours.

I think you're confusing the build environment and the packaged release.  I would say:

The build environment should meet standard expectations.  It should be flexible and work in a variety of environments.  I think you're trying way to hard to control things and I think that's going to backfire.

I think the key to usability is to check assumptions and generate nice error messages.  For the build, -std=c++17 should generate a nice error message if it is not supported:

```
$ clang -std=c++17 foo.cc
error: invalid value 'c++17' in '-std=c++17'
```

I'm not sure what other assumptions you have.  An autotools-like setup could help here.

When Hail starts up, it should compile a diagnostic example on the master and load it on the worker to verify everything is working.

As for the package, I'm all for including a compiler.

I'm admit, I'm a bit worried about deploying all of this.  As much as a design, we should have a deployment testing strategy.  I'm thinking something like:
 - When we have a merge candidate, have everyone on the team verify they can build and test in their development environment, possibly following some developer getting started instructions about new requirements.
 - Have a few people on the team spot-test the release packages.
 - Verify the cloud deployed versions work on Dataproc with cloudtools.  (This should be automated but isn't at the moment.   Something to add to the new CI.)
 - Have a few local users spot-test the release packages
 - Have a partner with a local Spark installation (e.g. GSK) test build and deploy.

Reply:
On the question of locking down compiler options, I realize most open-source software
build systems do their best to work with whatever CXXFLAGS you give them.

But ...

1. Hardly any of those open-source systems are trying to pull the trick we're playing,
    of dynamically-generating code which then gets compiled and linked in with the
    rest of the system.  That dramatically raises the stakes, because now a) the dynamic-compile
    compiler+options must produce code compatible with the build-system compiler and
    options; and b) if inherited options are used in the dynamic-compile, then any trouble
    they cause will manifest as an incomprehensible run-time failure seen by the end-user,
    not a build-time problem seen by a sysadmin or software guru.

2. It's often remarkably gnarly to pick up a large open-source package and get it to build on
    an arbitrary system, e.g. with a bleeding-edge compiler.   So much so that the Linux
    world has mostly gone over to repositories of prebuilt packages.  I think "Do it like other 
    open-source software build systems" is not setting the bar in the right place.
  
3. From long personal experience, it's fairly disastrous to get into the situation where the
    exact same source code on the exact same machine will build and work for one user, but
    not for another user.   If you're going to have software built by many people, then
    minimizing the variations in the build behavior will save effort.

Reply:
So I think I've been a bit vague about what I mean by "compatible" - the key constraint is
the dynamic-generated C++ and the src/main/c C++ files (currently in libhail.so) must be 
compiled for the same C++ ABI.

I don't think there's an issue for C sources - so maybe those should be split into a separate
library - and if you want those to inherit user compile options I guess I'd view it as mostly
harmless.

Now since there are only a small number of distinct not-completely-ancient g++ ABI standards,
*and* AFAIK clang conforms to (a subset of) those standards, then the "compatibility" issue
comes down to choosing whichever of those ABI's we can support on a particular machine -
which depends on what compiler we find.  And to make a prebuilt release which can work in a
wide variety of environments, we need to build the C++ library files (currently in libhail.so)
for each of the possible C++ ABI's (which I believe we can do using a recent g++ or clang,
but specifying the ABI version as an option.

So maybe we would ship with libhail.so (for C files), and two alternatives for the C++ files,
e.g. libhail_cpp_abi_v2.so and libhail_cpp_abi_v3.so.  And at first access to NativeBase,
we would load whichever was appropriate to match the compiler we're going to use for
dynamic-compile.

Then this would give us only a finite number of cases to test - possibly only 
linux old-abi, linux new-abi, and MacOS ?  Or maybe there are more abi's for MacOS as
well ?

I guess I'm ok with inheriting user CXXFLAGS for the library build and checking/warning about 
those which modify either the -std=c++XX, or the abi version (and enmforcing -fno-strict-align).
But I wouldn't want to inject arbitrary user compile options into the dynamic-compile, where
they might cause compilation to fail.

Reply:
> we need to build the C++ library files (currently in libhail.so) for each of the possible C++ ABI's

This seems way too complicated.  We're already shipping a compiler, let's just include libstdc++, too (and any other C++ ABI dependencies?  Are there any?)

 > But I wouldn't want to inject arbitrary user compile options into the dynamic-compile, where
they might cause compilation to fail.
 > Hardly any of those open-source systems are trying to pull the trick we're playing,
of dynamically-generating code which then gets compiled and linked in with the
rest of the system.

My feeling is, so what if the compilation fails?  I don't see any problem if we have a transparent error message about what we were trying to do and why it failed.

Is the goal here to make it never fail, or the goal here to make sure that "the situation where the exact same source code on the exact same machine will build and work for one user, but not for another user"?  It seems like you're invoking an extreme situation to argue for an altogether different one.

Anyway, I'm starting to sound like a broken record.  I've said my piece.  I'll let you and @danking settle the rest of the PR.  I realize if I don't like it I'm welcome to make the changes myself.